### PR TITLE
compiler: distinguish globals/imports/module-locals

### PR DESCRIPTION
**Pull request by josephsavona**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #29190
* #29189
* __->__ #29188

We currently use `LoadGlobal` and `StoreGlobal` to represent any read (or write) of a variable defined outside the component or hook that is being compiled. This is mostly fine, but for a lot of things we want to do going forward (resolving types across modules, for example) it helps to understand the actual source of a variable.

This PR is an incremental step in that direction. We continue to use LoadGlobal/StoreGlobal, but LoadGlobal now has a `binding:NonLocalBinding` instead of just the name of the global. The NonLocalBinding type tells us whether it was an import (and which kind, the source module name etc), a module-local binding, or a true global. By keeping the LoadGlobal/StoreGlobal instructions, most code that deals with "anything not declared locally" doesn't have to care about the difference. However, code that _does_ want to know the source of the value can figure it out.

### CODE REVIEW

File: compiler/packages/babel-plugin-react-compiler/src/HIR/HIRBuilder.ts [hunk 1]
```diff
@@ -253,17 +248,59 @@ export default class HIRBuilder {
...
-  ): Identifier | null {
+  ): VariableBinding {
     const originalName = path.node.name;
     const babelBinding = this.#resolveBabelBinding(path);
     if (babelBinding == null) {
-      return null;
+      return { kind: "Global", name: originalName };
+    }
+
+    // Check if the binding is from module scope, if so return null
```
Comment (mofeiZ) [inline compiler/packages/babel-plugin-react-compiler/src/HIR/HIRBuilder.ts:258]: Nit(comment): we're no longer returning null

File: compiler/packages/babel-plugin-react-compiler/src/HIR/HIRBuilder.ts [hunk 2]
```diff
@@ -212,12 +213,6 @@ export default class HIRBuilder {
     if (binding == null) {
       return null;
     }
-    // Check if the binding is from module scope, if so return null
-    const outerBinding =
-      this.parentFunction.scope.parent.getBinding(originalName);
-    if (binding === outerBinding) {
-      return null;
-    }
```
Comment (mofeiZ) [inline compiler/packages/babel-plugin-react-compiler/src/HIR/HIRBuilder.ts:220]: It looks like `#resolveBabelBinding` is also used by `isContextIdentifier`, should we also copy the module scope logic there?

File: compiler/packages/babel-plugin-react-compiler/src/HIR/HIR.ts [hunk 3]
```diff
@@ -1103,6 +1103,29 @@ export type MutableRange = {
   end: InstructionId;
 };
 
+export type VariableBinding =
+  // let, const, etc declared within the current component/hook
+  | { kind: "Identifier"; identifier: Identifier }
+  // bindings declard outside the current component/hook
+  | NonLocalBinding;
+
+export type NonLocalBinding =
```
Comment (mofeiZ) [inline compiler/packages/babel-plugin-react-compiler/src/HIR/HIR.ts:1112]: Niiice! Thanks for the descriptive comments

File: compiler/packages/babel-plugin-react-compiler/src/ReactiveScopes/MemoizeFbtOperandsInSameScope.ts [hunk 4]
```diff
@@ -67,7 +67,10 @@ function visit(fn: HIRFunction, fbtValues: Set<IdentifierId>): void {
          * all `fbt` string literals in case they are used as a jsx tag.
          */
         fbtValues.add(lvalue.identifier.id);
-      } else if (value.kind === "LoadGlobal" && FBT_TAGS.has(value.name)) {
+      } else if (
+        value.kind === "LoadGlobal" &&
+        FBT_TAGS.has(value.binding.name)
```
Comment (mofeiZ) [inline compiler/packages/babel-plugin-react-compiler/src/ReactiveScopes/MemoizeFbtOperandsInSameScope.ts:72]: I'm guessing that we want to keep this logic as-is (regardless of imports)? fbt-transform logic seems pretty fragile, so it seems fine to be conservative

**Review by mofeiZ — APPROVED** [review]

File: compiler/packages/babel-plugin-react-compiler/src/ReactiveScopes/MemoizeFbtOperandsInSameScope.ts (hunk 4 above)
Comment (josephsavona) [reply compiler/packages/babel-plugin-react-compiler/src/ReactiveScopes/MemoizeFbtOperandsInSameScope.ts:72]: yeah i'm leaving the fbt stuff alone, regardless of where fbt comes from. although the fbt plugin does require that you import it.

**Pull request merged**